File: ios/Cove/Flows/SendFlow/SendFlowHardwareScreen/SendFlowQrExport.swift
//
//  SendFlowQrExport.swift
//  Cove
//
//  Created by Praveen Perera on 11/24/24.
//
import SwiftUI

extension QrExportFormat: CaseIterable {
    public static var allCases: [QrExportFormat] { [.bbqr, .ur] }
}

struct SendFlowQrExport: View {
    let details: ConfirmDetails

    @State private var selectedFormat: QrExportFormat = .bbqr
    @State private var density: QrDensity = .init()
    @State private var qrs: [QrCodeView] = []
    @State private var error: String? = nil
    @State private var currentIndex = 0

    let startedAt: Date = .now

    /// Animation interval: dynamic based on density for both formats
    var animationInterval: TimeInterval {
        switch selectedFormat {
        case .bbqr: Double(density.bbqrAnimationIntervalMs()) / 1000.0
        case .ur: Double(density.urAnimationIntervalMs()) / 1000.0
        }
    }

    var body: some View {
        VStack {
            Text("Scan this QR")
                .font(.title3)
                .padding(.top, 12)
                .fontWeight(.semibold)

            Text("Scan with your hardware wallet\nto sign your transaction")
                .font(.callout)
                .foregroundStyle(.secondary)
                .multilineTextAlignment(.center)
                .padding(.top, 1)
                .padding(.horizontal, 40)

            Picker("Format", selection: $selectedFormat) {
                ForEach(QrExportFormat.allCases, id: \.self) { format in
                    Text(String(describing: format)).tag(format)
                }
            }
            .pickerStyle(.segmented)
            .padding(.vertical, 8)
            .frame(maxWidth: 200)

            QrContent
        }
        .onChange(of: selectedFormat) { _, _ in
            generateQrCodes()
        }
        .onChange(of: density) { _, _ in
            generateQrCodes()
        }
        .onAppear {
            generateQrCodes()
        }
    }

    @ViewBuilder
    var QrContent: some View {
        if let error {
            Text(error)
                .font(.footnote)
                .foregroundStyle(.red)
                .padding(.top, 8)
        } else if qrs.isEmpty {
            ProgressView()
                .padding(.top, 20)
        } else {
            AnimatedQrView
        }
    }

    @ViewBuilder
    var AnimatedQrView: some View {
        VStack {
            // .id() forces TimelineView recreation when interval changes
            TimelineView(.periodic(from: startedAt, by: animationInterval)) { context in
                let index = abs(Int(context.date.distance(to: startedAt) / animationInterval) % qrs.count)
                qrs[index]
                    .frame(maxWidth: .infinity)
                    .fixedSize(horizontal: false, vertical: true)
                    .padding(.horizontal, 11)
                    .onChange(of: index) { _, newIndex in
                        currentIndex = newIndex
                    }
            }
            .id(animationInterval)

            if qrs.count > 1 {
                HStack(alignment: .center, spacing: 8) {
                    MinusButtonMinimal
                    ProgressIndicator
                    PlusButtonMinimal
                }
                .padding(.horizontal, 9)
            } else {
                DensityButtons
                    .padding(.horizontal, 9)
            }
        }
    }

    @ViewBuilder
    var ProgressIndicator: some View {
        HStack(spacing: 4) {
            ForEach(0 ..< qrs.count, id: \.self) { index in
                Rectangle()
                    .fill(Color.blue)
                    .opacity(index == currentIndex ? 1 : 0.3)
                    .frame(height: 12)
                    .cornerRadius(2)
            }
        }
    }

    var canDecreaseDensity: Bool { density.canDecrease() }
    var canIncreaseDensity: Bool { density.canIncrease() && qrs.count > 1 }

    @ViewBuilder
    var MinusButtonMinimal: some View {
        Button { density = density.decrease() } label: {
            Image(systemName: "minus")
                .font(.system(size: 14, weight: .medium))
                .foregroundStyle(Color.secondary.opacity(canDecreaseDensity ? 1 : 0.3))
                .frame(width: 44, height: 44)
                .contentShape(Rectangle())
        }
        .disabled(!canDecreaseDensity)
    }

    @ViewBuilder
    var PlusButtonMinimal: some View {
        Button { density = density.increase() } label: {
            Image(systemName: "plus")
                .font(.system(size: 14, weight: .medium))
                .foregroundStyle(Color.secondary.opacity(canIncreaseDensity ? 1 : 0.3))
                .frame(width: 44, height: 44)
                .contentShape(Rectangle())
        }
        .disabled(!canIncreaseDensity)
    }

    @ViewBuilder
    var DensityButtons: some View {
        HStack(spacing: 0) {
            Button { density = density.decrease() } label: {
                Image(systemName: "minus")
                    .font(.system(size: 14, weight: .medium))
                    .frame(width: 32, height: 32)
                    .foregroundStyle(Color.secondary.opacity(canDecreaseDensity ? 1 : 0.3))
            }
            .disabled(!canDecreaseDensity)

            Divider()
                .frame(height: 20)

            Button { density = density.increase() } label: {
                Image(systemName: "plus")
                    .font(.system(size: 14, weight: .medium))
                    .frame(width: 32, height: 32)
                    .foregroundStyle(Color.secondary.opacity(canIncreaseDensity ? 1 : 0.3))
            }
            .disabled(!canIncreaseDensity)
        }
        .background(Color.secondary.opacity(0.15))
        .cornerRadius(50)
    }

    func generateQrCodes() {
        do {
            let strings: [String] = switch selectedFormat {
            case .bbqr:
                try details.psbtToBbqrWithDensity(density: density)
            case .ur:
                try details.psbtToUrWithDensity(density: density)
            }
            qrs = strings.map { QrCodeView(text: $0) }
            currentIndex = 0
            error = nil
        } catch let err {
            error = err.localizedDescription
            qrs = []
        }
    }
}

#Preview {
    AsyncPreview {
        SendFlowQrExport(details: confirmDetailsPreviewNew())
            .padding()
    }
}

#Preview("Sheet - Multi QR") {
    struct SheetPreview: View {
        @State private var isPresented = true

        var body: some View {
            VStack {
                Button("Show Sheet") {
                    isPresented = true
                }
            }
            .frame(maxWidth: .infinity, maxHeight: .infinity)
            .background(Color.midnightBlue.edgesIgnoringSafeArea(.all))
            .sheet(isPresented: $isPresented) {
                SendFlowQrExport(details: confirmDetailsPreviewNew())
                    .presentationDetents([.height(550), .height(650), .large])
                    .padding()
                    .padding(.top, 10)
            }
        }
    }

    return AsyncPreview {
        SheetPreview()
    }
}

File: android/app/src/main/java/org/bitcoinppl/cove/flows/SendFlow/SendFlowQrExport.kt
package org.bitcoinppl.cove.flows.SendFlow

import androidx.compose.foundation.Image
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.aspectRatio
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Add
import androidx.compose.material.icons.filled.Remove
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.SegmentedButton
import androidx.compose.material3.SegmentedButtonDefaults
import androidx.compose.material3.SingleChoiceSegmentedButtonRow
import androidx.compose.material3.Text
import androidx.compose.material3.VerticalDivider
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.asImageBitmap
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import kotlinx.coroutines.delay
import org.bitcoinppl.cove.QrCodeGenerator
import org.bitcoinppl.cove_core.types.ConfirmDetails
import org.bitcoinppl.cove_core.types.QrDensity
import org.bitcoinppl.cove_core.types.QrExportFormat

@Composable
fun SendFlowQrExport(
    details: ConfirmDetails,
    modifier: Modifier = Modifier,
) {
    var selectedFormat by remember { mutableStateOf(QrExportFormat.BBQR) }
    var density by remember { mutableStateOf(QrDensity()) }
    var qrStrings by remember { mutableStateOf<List<String>>(emptyList()) }
    var currentIndex by remember { mutableIntStateOf(0) }
    var error by remember { mutableStateOf<String?>(null) }

    fun generateQrCodes() {
        try {
            qrStrings =
                when (selectedFormat) {
                    QrExportFormat.BBQR -> details.psbtToBbqrWithDensity(density)
                    QrExportFormat.UR -> details.psbtToUrWithDensity(density)
                }
            error = null
            currentIndex = 0
        } catch (e: Exception) {
            error = e.message ?: "Unknown error"
            qrStrings = emptyList()
        }
    }

    // generate QR codes on initial load and when format/density changes
    LaunchedEffect(selectedFormat, density) {
        generateQrCodes()
    }

    // animation interval: dynamic based on density for both formats
    val animationDelayMs =
        when (selectedFormat) {
            QrExportFormat.BBQR -> density.bbqrAnimationIntervalMs().toLong()
            QrExportFormat.UR -> density.urAnimationIntervalMs().toLong()
        }

    // animate QR code cycling
    LaunchedEffect(qrStrings, animationDelayMs) {
        if (qrStrings.size > 1) {
            while (true) {
                delay(animationDelayMs)
                currentIndex = (currentIndex + 1) % qrStrings.size
            }
        }
    }

    Column(
        modifier = modifier,
        horizontalAlignment = Alignment.CenterHorizontally,
    ) {
        Text(
            text = "Scan this QR",
            style = MaterialTheme.typography.titleMedium,
            fontWeight = FontWeight.SemiBold,
            modifier = Modifier.padding(top = 12.dp),
        )

        Text(
            text = "Scan with your hardware wallet\nto sign your transaction",
            style = MaterialTheme.typography.bodySmall,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
            textAlign = TextAlign.Center,
            modifier = Modifier.padding(top = 4.dp, start = 40.dp, end = 40.dp),
        )

        // format picker (BBQR / UR)
        SingleChoiceSegmentedButtonRow(
            modifier =
                Modifier
                    .padding(vertical = 8.dp)
                    .width(200.dp),
        ) {
            QrExportFormat.entries.forEachIndexed { index, format ->
                SegmentedButton(
                    shape =
                        SegmentedButtonDefaults.itemShape(
                            index = index,
                            count = QrExportFormat.entries.size,
                        ),
                    onClick = { selectedFormat = format },
                    selected = selectedFormat == format,
                    label = { Text(format.toString()) },
                )
            }
        }

        // QR content
        if (error != null) {
            Text(
                text = error!!,
                style = MaterialTheme.typography.bodySmall,
                color = MaterialTheme.colorScheme.error,
                modifier = Modifier.padding(top = 8.dp),
            )
        } else if (qrStrings.isEmpty()) {
            Box(
                modifier =
                    Modifier
                        .fillMaxWidth()
                        .aspectRatio(1f)
                        .padding(horizontal = 11.dp),
                contentAlignment = Alignment.Center,
            ) {
                Text("Loading...")
            }
        } else {
            // animated QR view
            Column(horizontalAlignment = Alignment.CenterHorizontally) {
                Box(
                    modifier =
                        Modifier
                            .fillMaxWidth()
                            .aspectRatio(1f)
                            .padding(horizontal = 11.dp),
                    contentAlignment = Alignment.Center,
                ) {
                    val safeIndex = currentIndex.coerceIn(0, qrStrings.lastIndex.coerceAtLeast(0))
                    val qrString = qrStrings.getOrNull(safeIndex) ?: qrStrings.firstOrNull() ?: ""
                    val bitmap =
                        remember(qrString) {
                            QrCodeGenerator.generate(qrString, size = 512)
                        }

                    Image(
                        bitmap = bitmap.asImageBitmap(),
                        contentDescription = "QR code for transaction signing",
                        modifier = Modifier.fillMaxWidth(),
                    )
                }

                // density buttons and progress indicator
                if (qrStrings.size > 1) {
                    Row(
                        modifier = Modifier.padding(horizontal = 9.dp),
                        verticalAlignment = Alignment.CenterVertically,
                        horizontalArrangement = Arrangement.spacedBy(8.dp),
                    ) {
                        MinusButton(
                            enabled = density.canDecrease(),
                            onClick = { density = density.decrease() },
                        )

                        // progress indicator
                        Row(
                            modifier = Modifier.weight(1f),
                            horizontalArrangement = Arrangement.spacedBy(4.dp),
                        ) {
                            qrStrings.indices.forEach { index ->
                                Box(
                                    modifier =
                                        Modifier
                                            .weight(1f)
                                            .height(12.dp)
                                            .background(
                                                color =
                                                    MaterialTheme.colorScheme.primary.copy(
                                                        alpha = if (index == currentIndex) 1f else 0.3f,
                                                    ),
                                                shape = RoundedCornerShape(2.dp),
                                            ),
                                )
                            }
                        }

                        PlusButton(
                            enabled = density.canIncrease() && qrStrings.size > 1,
                            onClick = { density = density.increase() },
                        )
                    }
                } else {
                    // single QR - show combined density buttons
                    DensityButtons(
                        canDecrease = density.canDecrease(),
                        canIncrease = density.canIncrease(),
                        onDecrease = { density = density.decrease() },
                        onIncrease = { density = density.increase() },
                        modifier = Modifier.padding(horizontal = 9.dp),
                    )
                }
            }
        }
    }
}

@Composable
private fun MinusButton(
    enabled: Boolean,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
) {
    Box(
        modifier =
            modifier
                .size(44.dp)
                .clip(RoundedCornerShape(8.dp))
                .clickable(enabled = enabled, onClick = onClick),
        contentAlignment = Alignment.Center,
    ) {
        Icon(
            imageVector = Icons.Default.Remove,
            contentDescription = "Decrease density",
            tint =
                MaterialTheme.colorScheme.onSurfaceVariant.copy(
                    alpha = if (enabled) 1f else 0.3f,
                ),
            modifier = Modifier.size(18.dp),
        )
    }
}

@Composable
private fun PlusButton(
    enabled: Boolean,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
) {
    Box(
        modifier =
            modifier
                .size(44.dp)
                .clip(RoundedCornerShape(8.dp))
                .clickable(enabled = enabled, onClick = onClick),
        contentAlignment = Alignment.Center,
    ) {
        Icon(
            imageVector = Icons.Default.Add,
            contentDescription = "Increase density",
            tint =
                MaterialTheme.colorScheme.onSurfaceVariant.copy(
                    alpha = if (enabled) 1f else 0.3f,
                ),
            modifier = Modifier.size(18.dp),
        )
    }
}

@Composable
private fun DensityButtons(
    canDecrease: Boolean,
    canIncrease: Boolean,
    onDecrease: () -> Unit,
    onIncrease: () -> Unit,
    modifier: Modifier = Modifier,
) {
    Row(
        modifier =
            modifier
                .clip(RoundedCornerShape(50))
                .background(MaterialTheme.colorScheme.onSurface.copy(alpha = 0.15f)),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Box(
            modifier =
                Modifier
                    .size(32.dp)
                    .clip(RoundedCornerShape(50))
                    .clickable(enabled = canDecrease, onClick = onDecrease),
            contentAlignment = Alignment.Center,
        ) {
            Icon(
                imageVector = Icons.Default.Remove,
                contentDescription = "Decrease density",
                tint =
                    MaterialTheme.colorScheme.onSurfaceVariant.copy(
                        alpha = if (canDecrease) 1f else 0.3f,
                    ),
                modifier = Modifier.size(14.dp),
            )
        }

        VerticalDivider(
            modifier = Modifier.height(20.dp),
            color = MaterialTheme.colorScheme.onSurface.copy(alpha = 0.3f),
        )

        Box(
            modifier =
                Modifier
                    .size(32.dp)
                    .clip(RoundedCornerShape(50))
                    .clickable(enabled = canIncrease, onClick = onIncrease),
            contentAlignment = Alignment.Center,
        ) {
            Icon(
                imageVector = Icons.Default.Add,
                contentDescription = "Increase density",
                tint =
                    MaterialTheme.colorScheme.onSurfaceVariant.copy(
                        alpha = if (canIncrease) 1f else 0.3f,
                    ),
                modifier = Modifier.size(14.dp),
            )
        }
    }
}
